### DISCUSSION TITLE
STM32F411 ulora pin allocation problem

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#12759] [STM32 / Pyboard] [answered]
I have ulora running on both a ESP32 and a PICO and thought I'd try a Blackpill. So, I built the general version using the instructions here:
https://github.com/nspsck/STM32F411CEU6_BlackPill_Micropython/tree/main

However, I am having trouble setting-up the SPI.  Could someone point to a Blackpill example?  I am happy to use the SPI pins used in that build, ie I am not trying to change them.  The lines that I need to modify are:
```py
# baud rate to 5MHz
#        self.spi = SPI(self._spi_channel[0], baudrate=5000000,
#                       sck=Pin(self._spi_channel[1]), mosi=Pin(self._spi_channel[2]), miso=Pin(self._spi_channel[3]))
```
sck/mosi/miso are fixed in pin.csv so I don't think they need to be included.  The error I get is `mode is required`.  Does `SPI.CONTROLLER` need to be in there?

How do I find the required syntax for the SPI on the Blackpill?

Thanks.

**Comment by rkompass** [comment, marked as answer]
The Blackpill should be **very** similar to the Pyboard.
So iirc you have two options: use the pyb module spi or the machine module spi. For some reason I had to use SoftSPI in my works.
I seem to remember I had the same problem then. So this might be worth exploring, even on the Pyboard.

Could you change your code example to include imports and replace class references like `self._spi_channel[1]` with values that are self-consistent. Then we may try it out without reworking and guessing the values.

**Reply by davefes** [reply]
You have given me a few clues.  I thought there was only SPI in pyb.  I use SoftSPI on the ESP32, so will look at that.  I didn't know which part of ulora.py to include.  Here is a link to that file:
https://github.com/martynwheeler/u-lora/blob/main/ulora.py

I will look at SoftSPI.

Thank you.

**Reply by davefes** [reply]
Success, modified about line 132 in ulora.py:
```py
    self.spi = SoftSPI(baudrate=187500, polarity=1, phase=0, sck=Pin('B13'), mosi=Pin('B15'), miso=Pin('B14'))
```
Thanks for the prompts.
